packages/dashboard/lib/DashboardServer.ts

```diff
@@ -95,6 +98,62 @@ export class DashboardServer {
       this.expressApp.post("/rpc", this.postRpc.bind(this));
     }
 
+    this.expressApp.get("/fetch-and-compile", async (req, res) => {
+      const { address, networkId, etherscanApiKey } = req.query as Record<
+        string,
+        string
+      >;
+      let config;
+      try {
+        config = Config.detect();
+      } catch (error) {
+        const notFound = "Could not find suitable configuration file.";
+        if (!error.message.includes(notFound)) {
+          throw error;
+        }
+      }
+
+      // a key provided in the browser takes precedence over on in the config
+      let etherscanKey: undefined | string;
+      if (etherscanApiKey) {
+        etherscanKey = etherscanApiKey;
+      } else if (config && config.etherscan !== undefined) {
+        etherscanKey = config.etherscan.apiKey;
+      }
+
+      config = Config.default().merge({
+        networks: {
+          custom: { network_id: networkId }
+        },
+        network: "custom",
+        etherscan: {
+          apiKey: etherscanKey
+        }
+      });
+
+      let result;
+      try {
+        result = (await fetchAndCompile(address, config)).compileResult;
+      } catch (error) {
+        if (!error.message.includes("No verified sources")) {
+          throw error;
+        }
+      }
+      if (result) {
+        // we calculate hashes on the server because it is at times too
+        // resource intensive for the browser and causes it to crash
+        const hashes = result.compilations.map((compilation: Compilation) => {
+          return sha1(compilation);
+        });
+        res.json({
+          hashes,
+          compilations: result.compilations
+        });
+      } else {
+        res.json({ compilations: [] });
+      }
+    });
+
```

Comment:
If you open this as a separate PR, we can approve+merge it real fast and then this PR's diff will be like 60 lines easier to read!

Reply:
I don't understand, why would you pull this particular code out into its own PR?

Reply:
Ok, I can do that. If it is helpful I can give people an overview-type tour of this PR. Just let me know!

Reply:
Pulled this stuff out into https://github.com/trufflesuite/truffle/pull/6075